File: benchmarks/krabmaga/src/models/boids.rs
//! Boids on a continuous torus, updated synchronously.
//!
//! Henad's rule rather than Reynolds's, and rather than the one krABMaga's own `flockers` example
//! implements. Separation accumulates the offsets to everything inside the protected range,
//! alignment and cohesion average over everything inside the visual range, and the resulting
//! velocity is clamped into a speed band. Velocity is the displacement, so a tick is one unit of
//! time.
//!
//! Shape follows `flockers`: one agent per bird holding its own position and velocity, reading the
//! `Field2D` read buffer and re-registering itself with `set_object_location`. Nothing an agent
//! writes is visible until `Schedule::step` swaps the buffers, which is what makes the tick
//! synchronous.

use core::fmt;
use std::any::Any;
use std::hash::{Hash, Hasher};

use krabmaga::engine::agent::Agent;
use krabmaga::engine::fields::field::Field;
use krabmaga::engine::fields::field_2d::{toroidal_transform, Field2D, Location2D};
use krabmaga::engine::location::Real2D;
use krabmaga::engine::schedule::Schedule;
use krabmaga::engine::state::State;
use krabmaga::rand::{Rng, SeedableRng};
use krabmaga::rand_pcg::Pcg64;

#[derive(Clone, Copy)]
pub struct Boid {
    pub id: u32,
    pub loc: Real2D,
    pub vel: Real2D,
}

impl Location2D<Real2D> for Boid {
    fn get_location(self) -> Real2D {
        self.loc
    }

    fn set_location(&mut self, loc: Real2D) {
        self.loc = loc;
    }
}

impl Hash for Boid {
    fn hash<H: Hasher>(&self, hasher: &mut H) {
        self.id.hash(hasher);
    }
}

impl Eq for Boid {}

impl PartialEq for Boid {
    fn eq(&self, other: &Boid) -> bool {
        self.id == other.id
    }
}

impl fmt::Display for Boid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} loc {}", self.id, self.loc)
    }
}

impl Agent for Boid {
    fn step(&mut self, state: &mut dyn State) {
        let model = state.as_any().downcast_ref::<Boids>().expect("boids state");

        let mut close = (0.0f32, 0.0f32);
        let mut summed_velocity = (0.0f32, 0.0f32);
        let mut summed_offset = (0.0f32, 0.0f32);
        let mut seen = 0u32;

        let (shifts, count) = model.seam_shifts(self.loc);
        for &(sx, sy) in &shifts[..count] {
            let probe = Real2D {
                x: self.loc.x + sx,
                y: self.loc.y + sy,
            };
            for other in model
                .field
                .get_neighbors_within_relax_distance(probe, model.visual_range)
            {
                if other.id == self.id {
                    continue;
                }
                // Offset from the probe, which is the wrapped offset from this boid. Only one
                // probe can land a given neighbour inside the visual range, see `seam_shifts`.
                let dx = other.loc.x - probe.x;
                let dy = other.loc.y - probe.y;
                let dist_sq = dx * dx + dy * dy;

                if dist_sq < model.protected_sq {
                    close.0 -= dx;
                    close.1 -= dy;
                }
                if dist_sq < model.visual_sq {
                    summed_velocity.0 += other.vel.x;
                    summed_velocity.1 += other.vel.y;
                    summed_offset.0 += dx;
                    summed_offset.1 += dy;
                    seen += 1;
                }
            }
        }

        let mut vx = self.vel.x + close.0 * model.separation;
        let mut vy = self.vel.y + close.1 * model.separation;
        if seen > 0 {
            let inverse = 1.0 / seen as f32;
            vx += (summed_velocity.0 * inverse - self.vel.x) * model.alignment
                + summed_offset.0 * inverse * model.cohesion;
            vy += (summed_velocity.1 * inverse - self.vel.y) * model.alignment
                + summed_offset.1 * inverse * model.cohesion;
        }

        let speed_sq = vx * vx + vy * vy;
        if speed_sq > 0.0 {
            let speed = speed_sq.sqrt();
            if speed > model.max_speed {
                vx = vx / speed * model.max_speed;
                vy = vy / speed * model.max_speed;
            } else if speed < model.min_speed {
                vx = vx / speed * model.min_speed;
                vy = vy / speed * model.min_speed;
            }
        } else {
            vx = model.min_speed;
            vy = 0.0;
        }

        self.vel = Real2D { x: vx, y: vy };
        self.loc = Real2D {
            x: toroidal_transform(self.loc.x + vx, model.world_w),
            y: toroidal_transform(self.loc.y + vy, model.world_h),
        };
        model.field.set_object_location(*self, self.loc);
    }
}

pub struct Boids {
    pub world_w: f32,
    pub world_h: f32,
    visual_range: f32,
    visual_sq: f32,
    protected_sq: f32,
    separation: f32,
    alignment: f32,
    cohesion: f32,
    max_speed: f32,
    min_speed: f32,
    field: Field2D<Boid>,
    num_agents: u32,
    /// An exact `(x, y, vx, vy)` list for a gate scenario, in place of the random scatter.
    agents: Option<Vec<(f32, f32, f32, f32)>>,
    seed: u64,
}

#[allow(clippy::too_many_arguments)]
impl Boids {
    pub fn new(
        num_agents: u32,
        world_w: f32,
        world_h: f32,
        visual_range: f32,
        protected_range: f32,
        separation: f32,
        alignment: f32,
        cohesion: f32,
        max_speed: f32,
        min_speed: f32,
        seed: u64,
    ) -> Result<Self, String> {
        if visual_range <= 0.0 {
            return Err("visual_range must be positive".to_owned());
        }
        // The seam probes below assume a neighbourhood narrower than half the world. See
        // `seam_shifts`.
        if 2.0 * visual_range > world_w.min(world_h) {
            return Err(format!(
                "visual_range {visual_range} needs a world at least {} across, got {world_w} by {world_h}",
                2.0 * visual_range
            ));
        }
        Ok(Boids {
            world_w,
            world_h,
            visual_range,
            visual_sq: visual_range * visual_range,
            protected_sq: protected_range * protected_range,
            separation,
            alignment,
            cohesion,
            max_speed,
            min_speed,
            // One bucket per visual range. Coarser and the field's own bucket scan stops short of
            // the radius, finer and it never reaches it.
            field: Field2D::new(world_w, world_h, visual_range, true),
            num_agents,
            agents: None,
            seed,
        })
    }

    pub fn with_agents(mut self, agents: &[(f32, f32, f32, f32)]) -> Self {
        self.num_agents = agents.len() as u32;
        self.agents = Some(agents.to_vec());
        self
    }

    pub fn population(&self) -> u64 {
        u64::from(self.num_agents)
    }

    /// Where to stand when asking the field for neighbours.
    ///
    /// `get_neighbors_within_*_distance` clamps its bucket scan to the field on a toroidal world,
    /// so it never visits the buckets across a seam. Asking again from the mirrored point one world
    /// away reaches them. A neighbour lands inside the visual range of exactly one probe, since the
    /// probes are a whole world apart and the range is under half of that, so nothing is counted
    /// twice and the offset from the probe is already the wrapped one.
    fn seam_shifts(&self, loc: Real2D) -> ([(f32, f32); 4], usize) {
        let mut xs = [0.0f32; 2];
        let mut nx = 1;
        if loc.x <= self.visual_range {
            xs[1] = self.world_w;
            nx = 2;
        } else if loc.x >= self.world_w - self.visual_range {
            xs[1] = -self.world_w;
            nx = 2;
        }

        let mut ys = [0.0f32; 2];
        let mut ny = 1;
        if loc.y <= self.visual_range {
            ys[1] = self.world_h;
            ny = 2;
        } else if loc.y >= self.world_h - self.visual_range {
            ys[1] = -self.world_h;
            ny = 2;
        }

        let mut shifts = [(0.0f32, 0.0f32); 4];
        let mut count = 0;
        for &x in &xs[..nx] {
            for &y in &ys[..ny] {
                shifts[count] = (x, y);
                count += 1;
            }
        }
        (shifts, count)
    }
}

impl State for Boids {
    fn init(&mut self, schedule: &mut Schedule) {
        let mut rng = Pcg64::seed_from_u64(self.seed);
        let speed = 0.5 * (self.min_speed + self.max_speed);

        for id in 0..self.num_agents {
            let (x, y, vx, vy) = match &self.agents {
                Some(agents) => agents[id as usize],
                None => {
                    // Scattered at random with a fixed speed and a random heading, as Henad does.
                    let x = rng.random::<f32>() * self.world_w;
                    let y = rng.random::<f32>() * self.world_h;
                    let angle = rng.random::<f32>() * std::f32::consts::TAU;
                    (x, y, angle.cos() * speed, angle.sin() * speed)
                }
            };
            let boid = Boid {
                id,
                loc: Real2D { x, y },
                vel: Real2D { x: vx, y: vy },
            };
            self.field.set_object_location(boid, boid.loc);
            schedule.schedule_repeating(Box::new(boid), 0.0, 0);
        }
    }

    fn update(&mut self, _step: u64) {
        self.field.lazy_update();
    }

    fn reset(&mut self) {
        self.field = Field2D::new(self.world_w, self.world_h, self.visual_range, true);
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn as_state(&self) -> &dyn State {
        self
    }

    fn as_state_mut(&mut self) -> &mut dyn State {
        self
    }
}

/// `x y vx vy` per agent, in creation order, as the fixture format wants.
///
/// The schedule hands its agents back in hash order, hence the sort.
pub fn rows(schedule: &Schedule) -> Vec<(f32, f32, f32, f32)> {
    let mut boids: Vec<Boid> = schedule
        .get_all_events()
        .iter()
        .filter_map(|agent| agent.downcast_ref::<Boid>().copied())
        .collect();
    boids.sort_by_key(|boid| boid.id);
    boids
        .iter()
        .map(|boid| (boid.loc.x, boid.loc.y, boid.vel.x, boid.vel.y))
        .collect()
}

File: benchmarks/mason/Boids.java
import sim.engine.Schedule;
import sim.engine.SimState;
import sim.engine.Steppable;
import sim.field.continuous.Continuous2D;
import sim.util.Bag;
import sim.util.Double2D;

/**
 * Boids on a continuous torus, updated synchronously.
 *
 * Henad's rule rather than Reynolds's, and rather than the one `sim.app.flockers` implements.
 * Separation accumulates the offsets to everything inside the protected range, alignment and
 * cohesion average over everything inside the visual range, and the resulting velocity is clamped
 * into a speed band. Velocity is the displacement, so a tick is one unit of time.
 *
 * The scaffolding follows flockers'. A `Continuous2D`, one Steppable per boid, and a cached
 * `Double2D` per boid rather than a lookup on the field. MASON shuffles the boids sharing an
 * ordering, so the step is split across two of them. Every boid computes at ordering 0, and one
 * commit Steppable at ordering 1 moves them all.
 */
public class Boids extends HenadModel {
    /** MASON's own continuous-space examples size their buckets this way. */
    private static final double DISCRETIZATION_FACTOR = 1.5;

    /**
     * Bucket size for the neighbour index, dividing both extents exactly.
     *
     * MASON wraps bucket indices rather than coordinates, and a bucket grid that does not tile the
     * torus loses neighbours across the seam. Neither query reports it. Brute force against an
     * O(n^2) scan finds no misses once both extents divide.
     */
    private static double discretization(double width, double height, double range) {
        final double target = Math.max(range / DISCRETIZATION_FACTOR, 1e-9);
        for (int columns = Math.max(1, (int) (width / target)); columns >= 1; columns--) {
            final double cell = width / columns;
            final double rows = height / cell;
            if (Math.abs(rows - Math.rint(rows)) < 1e-9 * Math.max(1.0, rows)) {
                return cell;
            }
        }
        // One bucket per axis for a world no cell tiles. Correct, and every query returns everything.
        return Math.max(width, height);
    }

    static final class Boid implements Steppable {
        Double2D loc;
        double vx;
        double vy;
        Double2D nextLoc;
        double nextVx;
        double nextVy;

        Boid(double x, double y, double vx, double vy) {
            this.loc = new Double2D(x, y);
            this.vx = vx;
            this.vy = vy;
            this.nextLoc = this.loc;
            this.nextVx = vx;
            this.nextVy = vy;
        }

        public void step(SimState state) {
            final Boids m = (Boids) state;
            final Continuous2D space = m.space;
            final double x = loc.x;
            final double y = loc.y;
            // The candidate query, filtered exactly below. MASON's exact one is inclusive at the
            // radius where the rule is strict, and it costs a lookup on the field per candidate.
            final Bag near = space.getNeighborsWithinDistance(loc, m.visualRange, true, false, m.candidates);

            double closeX = 0.0;
            double closeY = 0.0;
            double sumVx = 0.0;
            double sumVy = 0.0;
            double sumDx = 0.0;
            double sumDy = 0.0;
            int seen = 0;

            for (int i = 0; i < near.numObjs; i++) {
                final Boid other = (Boid) near.objs[i];
                if (other == this) {
                    continue;
                }
                // tdx(a, b) gives a - b the short way round, the offset the rule wants.
                final double dx = space.tdx(other.loc.x, x);
                final double dy = space.tdy(other.loc.y, y);
                final double distSq = dx * dx + dy * dy;
                if (distSq < m.protectedSq) {
                    closeX -= dx;
                    closeY -= dy;
                }
                if (distSq < m.visualSq) {
                    sumVx += other.vx;
                    sumVy += other.vy;
                    sumDx += dx;
                    sumDy += dy;
                    seen++;
                }
            }

            double newVx = vx + closeX * m.separation;
            double newVy = vy + closeY * m.separation;
            if (seen > 0) {
                final double inv = 1.0 / seen;
                newVx += (sumVx * inv - vx) * m.alignment + sumDx * inv * m.cohesion;
                newVy += (sumVy * inv - vy) * m.alignment + sumDy * inv * m.cohesion;
            }

            final double speed = Math.sqrt(newVx * newVx + newVy * newVy);
            if (speed > 0.0) {
                if (speed > m.maxSpeed) {
                    newVx = newVx / speed * m.maxSpeed;
                    newVy = newVy / speed * m.maxSpeed;
                } else if (speed < m.minSpeed) {
                    newVx = newVx / speed * m.minSpeed;
                    newVy = newVy / speed * m.minSpeed;
                }
            } else {
                newVx = m.minSpeed;
                newVy = 0.0;
            }

            nextVx = newVx;
            nextVy = newVy;
            nextLoc = new Double2D(space.tx(x + newVx), space.ty(y + newVy));
        }
    }

    private static final Steppable COMMIT = new Steppable() {
        public void step(SimState state) {
            final Boids m = (Boids) state;
            for (Boid boid : m.boids) {
                boid.loc = boid.nextLoc;
                boid.vx = boid.nextVx;
                boid.vy = boid.nextVy;
                m.space.setObjectLocation(boid, boid.loc);
            }
        }
    };

    public final Continuous2D space;
    final double visualRange;
    final double visualSq;
    final double protectedSq;
    final double separation;
    final double alignment;
    final double cohesion;
    final double maxSpeed;
    final double minSpeed;
    /** One Bag reused by every neighbour query, as MASON's docs ask. */
    final Bag candidates = new Bag();
    Boid[] boids;

    private final int numAgents;
    private final double worldWidth;
    private final double worldHeight;
    private final double[][] placed;

    /**
     * `placed` gives an exact `(x, y, vx, vy)` list for a gate scenario, otherwise the population
     * is scattered at random with a fixed speed and a random heading, as Henad does.
     */
    public Boids(long seed, int numAgents, double worldWidth, double worldHeight, double visualRange,
            double protectedRange, double separation, double alignment, double cohesion,
            double maxSpeed, double minSpeed, double[][] placed) {
        super(seed);
        this.numAgents = numAgents;
        this.worldWidth = worldWidth;
        this.worldHeight = worldHeight;
        this.visualRange = visualRange;
        this.visualSq = visualRange * visualRange;
        this.protectedSq = protectedRange * protectedRange;
        this.separation = separation;
        this.alignment = alignment;
        this.cohesion = cohesion;
        this.maxSpeed = maxSpeed;
        this.minSpeed = minSpeed;
        this.placed = placed;
        this.space = new Continuous2D(discretization(worldWidth, worldHeight, visualRange), worldWidth, worldHeight);
    }

    public void start() {
        super.start();
        space.clear();
        boids = placed != null ? fromTable() : scattered();
        for (Boid boid : boids) {
            space.setObjectLocation(boid, boid.loc);
            schedule.scheduleRepeating(Schedule.EPOCH, 0, boid, 1.0);
        }
        schedule.scheduleRepeating(Schedule.EPOCH, 1, COMMIT, 1.0);
    }

    private Boid[] fromTable() {
        final Boid[] made = new Boid[placed.length];
        for (int i = 0; i < placed.length; i++) {
            final double[] row = placed[i];
            made[i] = new Boid(row[0], row[1], row[2], row[3]);
        }
        return made;
    }

    private Boid[] scattered() {
        final Boid[] made = new Boid[numAgents];
        final double speed = 0.5 * (minSpeed + maxSpeed);
        for (int i = 0; i < numAgents; i++) {
            final double angle = random.nextDouble() * 2.0 * Math.PI;
            made[i] = new Boid(random.nextDouble() * worldWidth, random.nextDouble() * worldHeight,
                    Math.cos(angle) * speed, Math.sin(angle) * speed);
        }
        return made;
    }

    public int population() {
        return boids.length;
    }

    /** `x y vx vy` per boid, in creation order, as the fixture format wants. */
    public double[][] rows() {
        final double[][] out = new double[boids.length][];
        for (int i = 0; i < boids.length; i++) {
            final Boid boid = boids[i];
            out[i] = new double[] { boid.loc.x, boid.loc.y, boid.vx, boid.vy };
        }
        return out;
    }
}
